fendermint/vm/topdown/proof-service/src/service.rs

```diff
@@ -0,0 +1,323 @@
+                gateway_actor_id,
+                subnet_id.clone(),
+            )
+            .context("Failed to create proof assembler")?,
+        );
+
+        Ok(Self {
+            config,
+            cache,
+            f3_client,
+            assembler,
+        })
+    }
+
+    /// Main service loop - runs continuously and polls parent chain periodically
+    ///
+    /// Maintains a ticker that triggers proof generation at regular intervals.
+    /// Errors are logged but don't stop the service - it will retry on next tick.
+    pub async fn run(self) {
+        tracing::info!(
+            polling_interval = ?self.config.polling_interval,
+            lookahead = self.config.lookahead_instances,
+            "Starting proof generator service"
+        );
+
+        // Validator is already initialized in new() with trusted power table
+        let mut poll_interval = interval(self.config.polling_interval);
+        poll_interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+
+        loop {
+            poll_interval.tick().await;
+
+            tracing::debug!("Poll interval tick");
+            if let Err(e) = self.generate_next_proofs().await {
+                tracing::error!(
+                    error = %e,
+                    "Failed to generate proofs, will retry on next tick"
+                );
+            }
+        }
+    }
+
+    /// Generate proofs for next needed instances
+    ///
+    /// Called by run() on each tick. Implements the core flow:
+    /// FETCH → VALIDATE → GENERATE → CACHE
+    ///
+    /// CRITICAL: Processes F3 instances SEQUENTIALLY - never skips!
+    async fn generate_next_proofs(&self) -> Result<()> {
+        let last_committed = self.cache.last_committed_instance();
+        // Start FROM last_committed (not +1) - F3 client expects to validate from current state
+        let next_instance = last_committed;
+        let max_instance = last_committed + self.config.lookahead_instances;
+
+        tracing::debug!(
+            last_committed,
+            next_instance,
+            max_instance,
+            "Checking for new F3 certificates"
+        );
+
+        // Process instances IN ORDER - this is critical for F3
+        for instance_id in next_instance..=max_instance {
+            // Skip if already cached
+            if self.cache.contains(instance_id) {
+                tracing::debug!(instance_id, "Proof already cached");
+                continue;
+            }
+
+            // Check if F3 client has already passed this instance
+            {
+                let f3_client = self.f3_client.lock().await;
+                let f3_current = f3_client.current_instance();
+                if f3_current > instance_id {
+                    tracing::debug!(
+                        instance_id,
+                        f3_current,
+                        "F3 client already past this instance, skipping"
+                    );
+                    continue;
```

Comment:
In this case the last committed instance ID in the cache is not up-to-date with the chain state. May be worth updating it or somehow completely prevent this from happening...

Reply:
Yes. This is not hooked in this PR. As the chain interaction is completely missing here. Hence why it just blindly relies on storage.